File: golang/core/medium/min_number_pushes_on_keyboard_2/min_number_pushes_on_keyboard_2.go
package main

import (
	"fmt"
	"sort"
)

func minimumPushes(word string) int {
	hashmapKeyboard := map[string][]string{
		"2": {},
		"3": {},
		"4": {},
		"5": {},
		"6": {},
		"7": {},
		"8": {},
		"9": {},
	}

	hashmapWord := map[string]int{}
	for _, letter := range word {
		hashmapWord[string(letter)]++
	}

	// create sortedEntries and sort them
	type kv struct {
		Key   string
		Value int
	}
	var sortedEntries []kv
	for k, v := range hashmapWord {
		sortedEntries = append(sortedEntries, kv{k, v})
	}
	sort.Slice(sortedEntries, func(i, j int) bool {
		return sortedEntries[i].Value > sortedEntries[j].Value
	})


	keys := []string{}
	for k := range hashmapKeyboard {
		keys = append(keys, k)
	}
	sort.Strings(keys)

    index := 0
	for _, entry := range sortedEntries {
		numberKey := keys[index % len(keys)]
		hashmapKeyboard[numberKey] = append(hashmapKeyboard[numberKey], entry.Key)
		index++
	}

	reverseLookup := map[string]struct {
		number string
		index  int
	}{}
	for number, letters := range hashmapKeyboard {
		for idx, letter := range letters {
			reverseLookup[letter] = struct {
				number string
				index  int
			}{number, idx}
		}
	}

	countOfPushes := 0
	for _, letter := range word {
		info, found := reverseLookup[string(letter)]
		if found {
			countOfPushes += info.index + 1
		}
	}
	return countOfPushes
}

func main() {
	fmt.Println(minimumPushes("aabbccddeeffgghhiiiiii"))
}

File: typescript/core/medium/min_number_pushes_on_keyboard_2.ts
function minimumPushes(word: string): number {
    const hashmapKeyboard: Map<string, string[]> = new Map<string, string[]>([
        ['2', []],
        ['3', []],
        ['4', []],
        ['5', []],
        ['6', []],
        ['7', []],
        ['8', []],
        ['9', []]
    ])

    const hashmapWord: Map<string, number> = new Map<string, number>()
    for (const letter of word) {
        hashmapWord.set(letter, (hashmapWord.get(letter) || 0) + 1)
    }

    const sortedEntries = Array.from(hashmapWord.entries()).sort((a, b) => b[1] - a[1])
    const numbers = Array.from(hashmapKeyboard.keys())

    let index = 0
    for (const [letter, _] of sortedEntries) {
        const numberKey = numbers[index % numbers.length]
        hashmapKeyboard.get(numberKey)?.push(letter)
        index++
    }

    const reverseLookup: Map<string, { number: string, index: number }> = new Map()
    for (const [number, letters] of hashmapKeyboard) {
        letters.forEach((letter, idx) => {
            reverseLookup.set(letter, { number, index: idx })
        })
    }

    let countOfPushes = 0
    for (const letter of word) {
        const info = reverseLookup.get(letter)
        if (info) {
            countOfPushes += info.index + 1
        }
    }
    return countOfPushes
}

console.log(minimumPushes("aabbccddeeffgghhiiiiii"))
